pulp_rpm/app/tasks/publishing.py

```diff
@@ -156,28 +161,53 @@ def populate(self):
             self.handle_sub_repos(distribution_tree)
 
         all_content = main_content
-        for name, content in self.sub_repos:
+        for name, content, checksums in self.sub_repos:
             os.mkdir(name)
             setattr(self, f"{name}_content", content)
+            setattr(self, f"{name}_checksums", checksums)
             setattr(self, f"{name}_repomdrecords", self.prepare_metadata_files(content, name))
             all_content |= content
         self.publish_artifacts(all_content)
 
 
-def publish(repository_version_pk, metadata_signing_service=None):
+def get_checksum(name, checksums):
+    """
+    Get checksum algorithm for publishing metadata.
+
+    Args:
+        name (str): Name of the metadata type.
+        checksums (dict): Checksums for metadata and packages.
+
+    """
+    original = checksums.get("original")
+    metadata = checksums.get("metadata")
+    package = checksums.get("package")
+    checksum_type = original.get(name, CHECKSUM_TYPES.SHA256.upper())
+
+    if metadata and name not in PACKAGE_REPODATA:
```

Comment:
can you help me to understand this part?
package_repodata contains PACKAGE_REPODATA = ['primary', 'filelists', 'other'] how it can possibly container `metadata` if the `metadata` variable is checksum_type that has been specified during publish?

Reply:
with this approach i ended up having half or repodata with original sha256 and half with the sha512 i have specified.

I think we want to have all repodata with the same checksum, don't we?
```
<?xml version="1.0" encoding="UTF-8"?>
<repomd xmlns="http://linux.duke.edu/metadata/repo" xmlns:rpm="http://linux.duke.edu/metadata/rpm">
  <revision>1586347179</revision>
  <data type="primary">
    <checksum type="sha256">ca37b899e3bcbd127fce0897c9ce7b36081d23150baa0bfd880be86360e7cc69</checksum>
    <open-checksum type="sha256">a5aaa83fcac98bcc9a084edfda2dbd25d3d4751189a0fbd9e693c9a7d7ec38ac</open-checksum>
    <location href="repodata/ca37b899e3bcbd127fce0897c9ce7b36081d23150baa0bfd880be86360e7cc69-primary.xml.gz"/>
    <timestamp>1586347179</timestamp>
    <size>4092</size>
    <open-size>34974</open-size>
  </data>
  <data type="filelists">
    <checksum type="sha256">f015f92b021a6fe869a3bb1d8f1319f5239ada2656e8a6a5e5f56e7c3c93bb70</checksum>
    <open-checksum type="sha256">7897e43cf674bfd4744f1bf0de7d00e75f819090eb11969a557b77f022091503</open-checksum>
    <location href="repodata/f015f92b021a6fe869a3bb1d8f1319f5239ada2656e8a6a5e5f56e7c3c93bb70-filelists.xml.gz"/>
    <timestamp>1586347179</timestamp>
    <size>2108</size>
    <open-size>6912</open-size>
  </data>
  <data type="other">
    <checksum type="sha256">b906ec0dc073608e7ecc5890b950ba5803652da89326b244189aba5b10f22d43</checksum>
    <open-checksum type="sha256">866f4f8caa8fc4720a173670844b8f380c12523ed32cde20bfdf29d7d0627436</open-checksum>
    <location href="repodata/b906ec0dc073608e7ecc5890b950ba5803652da89326b244189aba5b10f22d43-other.xml.gz"/>
    <timestamp>1586347179</timestamp>
    <size>1931</size>
    <open-size>5830</open-size>
  </data>
  <data type="primary_db">
    <checksum type="sha512">ed4658fba564873ffdffc65569739d796220b07566f637ad25ecd0e227ab03faec3fe00de3ffc03f495c4f81f8046243b461e7593a81472edb586c3d2783ee6f</checksum>
    <open-checksum type="sha512">bd7d3d6bad4419fa55b4f3a2b4d0389d5aca06783f57ca48ddb362163b7eb71e90e37dc77523e9980d1623f425cca321ac104aede9795149a880c039a96e3ba2</open-checksum>
    <location href="repodata/ed4658fba564873ffdffc65569739d796220b07566f637ad25ecd0e227ab03faec3fe00de3ffc03f495c4f81f8046243b461e7593a81472edb586c3d2783ee6f-primary.sqlite.bz2"/>
    <timestamp>1586347179</timestamp>
    <size>7040</size>
    <open-size>46080</open-size>
    <database_version>0</database_version>
  </data>
  <data type="filelists_db">
    <checksum type="sha512">5dbd3518170379a660774a1168b5091c6bb2d0ba669a9d89400c9bb7a75b5283b1e361e0521b4c3d3fe69ea7152d266b44c79f1c2ea847e9ce1d95b184ce656f</checksum>
    <open-checksum type="sha512">7cbb81811edddea3ca3a1db958c92bfe1dc8eb01385feb9915725b19b0b23ac79e7e2c1358f48f6eadabe3bb678cf5c37f576672bb55597ed74b90bee8eb35d5</open-checksum>
    <location href="repodata/5dbd3518170379a660774a1168b5091c6bb2d0ba669a9d89400c9bb7a75b5283b1e361e0521b4c3d3fe69ea7152d266b44c79f1c2ea847e9ce1d95b184ce656f-filelists.sqlite.bz2"/>
    <timestamp>1586347179</timestamp>
    <size>3633</size>
    <open-size>13312</open-size>
    <database_version>0</database_version>
  </data>
  <data type="other_db">
    <checksum type="sha512">83d9a7f28804fb641ce19b3770693d315e0a2373cffd3726535ee5e5c27ae4920a2ee4b95d397c5781a153f5bd4e38de5627aa7a5cd7301fb460ef4cf752240c</checksum>
    <open-checksum type="sha512">6ae0aa2c6d642761724e7636c81013697be3c8ee8ffff08d951ce533fb0dc9983432df17e084a1ceaf79e7ca35aa0a48ee63eb74f53ff4e7164d65f125f82504</open-checksum>
    <location href="repodata/83d9a7f28804fb641ce19b3770693d315e0a2373cffd3726535ee5e5c27ae4920a2ee4b95d397c5781a153f5bd4e38de5627aa7a5cd7301fb460ef4cf752240c-other.sqlite.bz2"/>
    <timestamp>1586347179</timestamp>
    <size>2744</size>
    <open-size>12288</open-size>
    <database_version>0</database_version>
  </data>
  <data type="updateinfo">
    <checksum type="sha512">229cae19d5a7066ab1c3b1e79b6611b2330e05449fa9af6f8a3f91d8c10754c85afde060023c8f9ba98d5f63cee0309e4df7f8181d09a0dcc584599a50c621b3</checksum>
    <open-checksum type="sha512">82156ae04ffe639bc16c3063aaf88f667af97f263ad7bf81468f6e37395761a12466ca91f53a389a56fce4041a54e24253fbb01fcc28f848570330ec39a95bde</open-checksum>
    <location href="repodata/229cae19d5a7066ab1c3b1e79b6611b2330e05449fa9af6f8a3f91d8c10754c85afde060023c8f9ba98d5f63cee0309e4df7f8181d09a0dcc584599a50c621b3-updateinfo.xml.gz"/>
    <timestamp>1586347179</timestamp>
    <size>778</size>
    <open-size>5613</open-size>
  </data>
  <data type="modules">
    <checksum type="sha512">a11ab11edae85bdefe30af94a00b1e735accbc1f2aa5b8b2fe125a0c4e8d47bb3b737f9e5a8f74f84fbfc5d4710103d6e0847448cbeee98f59f7eb91e5d9a711</checksum>
    <location href="repodata/a11ab11edae85bdefe30af94a00b1e735accbc1f2aa5b8b2fe125a0c4e8d47bb3b737f9e5a8f74f84fbfc5d4710103d6e0847448cbeee98f59f7eb91e5d9a711-modules.yaml"/>
    <timestamp>1586347179</timestamp>
    <size>3785</size>
  </data>
  <data type="group">
    <checksum type="sha512">d17afd785c891d70a57959a10670fd062f4ff974456f4ce1099c5445bdbe48e020606b82348e5f791a9ea8e19fd08f81964659b9657df5e881a708029ea48062</checksum>
    <location href="repodata/d17afd785c891d70a57959a10670fd062f4ff974456f4ce1099c5445bdbe48e020606b82348e5f791a9ea8e19fd08f81964659b9657df5e881a708029ea48062-comps.xml"/>
    <timestamp>1586347179</timestamp>
    <size>2222</size>
  </data>
</repomd>
```

Reply:
metadata is checksum_type, but I'm testing if it is None and also if the data_type is ['primary', 'filelists', 'other']

Reply:
about the second question, I really don't know, my understanding is:
* if `package_checksum_type` only change checksum_type of ['primary', 'filelists', 'other']
* if `metadata_checksum_type` change checksum_type except on ['primary', 'filelists', 'other']
This logic could cause different checksum types as you pointed.
@goosemania could you help us?

Reply:
`metadata_checksum_type` is a configuration for all metadata, basically everything in repomd. So everything listed in repomd should have the same checksum type. It would be either the `metadata_checksum_type` or, if not specified, SHA256. No other choice.

`package_checksum_type` is a configuration for packages listed in any metadata files.
Because of on_demand sync, it is possible that the requested checksum is not available. In such case, you need to choose one of the available checksums:
 - the requested one
 - if ^ is not available, then the one from the remote repo
 - if this data is not available for whatever reason, choose sha256.
Packages in metadata files might end up having different checksum types and that's ok.

Let me know if it doesn't help.

Reply:
let's say I do:
```python
metadata_checksum_type = sha256
package_checksum_type = sha512
```
what should be the output?